src/main/scala/org/apache/mesos/chronos/scheduler/jobs/Schedule.scala

```diff
@@ -1,44 +1,162 @@
-import org.joda.time.{DateTime, Period}
+import org.joda.time.{DateTimeZone, DateTime, Period}
+
+import scala.util.{Failure, Success, Try}
 
+/**
+ * Base trait for all schedule types
+ */
+sealed trait Schedule extends Product { self =>
 
-case class Schedule(@JsonProperty schedule: String,
-                    @JsonProperty scheduleTimeZone: String,
-                    @JsonProperty invocationTime: DateTime,
-                    @JsonProperty originTime: DateTime,
-                    @JsonProperty offset: Long = 0,
-                    @JsonProperty recurrences: Option[Long] = None,
-                    @JsonProperty period: Period) {
+  type SType <: Schedule { type SType = self.SType }
+
+  def schedule: String
+  def scheduleTimeZone: String
+  def scheduleType: ScheduleType
+  def invocationTime: DateTime
+  def recurrences: Option[Long]
+  def next: Option[SType]
+  def toStringRepresentation: String
+}
 
+/**
+ * A schedule that uses ISO 8601 repeating interval notation
+ *
+ * @param schedule the text representation of the schedule
+ * @param scheduleTimeZone the time zone the schedule should be used in
+ * @param invocationTime the time this schedule needs to be invoked (UTC)
+ * @param originTime the root time that the ISO 8601 schedule was started (UTC)
+ * @param offset the number of executions after the origin that have completed
+ * @param recurrences the number of remaining executions, or -1 if this is an infinite schedule
+ * @param period the time between executions
+ */
+case class Iso8601Schedule(@JsonProperty schedule: String,
+                           @JsonProperty scheduleTimeZone: String,
+                           @JsonProperty invocationTime: DateTime,
+                           @JsonProperty originTime: DateTime,
+                           @JsonProperty offset: Long = 0,
+                           @JsonProperty recurrences: Option[Long] = None,
+                           @JsonProperty period: Period) extends Schedule {
+  type SType = Iso8601Schedule
 
-  def next: Option[Schedule] = {
+  val scheduleType = Iso8601Type
+
+  override def next: Option[Iso8601Schedule] = {
     if (recurrences.exists(_ == 0)) None
     else {
       val nextOffset = offset + 1
-      val nextInvocationTime = Schedule.addPeriods(originTime, period, nextOffset.asInstanceOf[Int] /* jodatime doesn't support longs in their plus methods. What to do about overflow? */)
+      val nextInvocationTime = Schedules.addPeriods(originTime, period, nextOffset.asInstanceOf[Int] /* jodatime doesn't support longs in their plus methods. What to do about overflow? */)
 
-      Some(Schedule(schedule, scheduleTimeZone, nextInvocationTime, originTime, nextOffset, recurrences.map(_ - 1), period))
+      Some(Iso8601Schedule(schedule, scheduleTimeZone, nextInvocationTime, originTime, nextOffset, recurrences.map(_ - 1), period))
     }
   }
 
-  def toZeroOffsetISO8601Representation: String = {
+  def toStringRepresentation: String = {
```

Comment:
Could we just override `toString` here? It is not clear how `toStringRepresentation` is different from conventional `toString`.